services/device-registry-base/src/main/java/org/eclipse/hono/deviceregistry/service/deviceconnection/MapBasedDeviceConnectionService.java

```diff
@@ -128,9 +129,26 @@ public MapBasedDeviceConnectionsConfigProperties getConfig() {
         final int currentMapSize = adapterInstancesForTenantMap.size();
         if (currentMapSize < getConfig().getMaxDevicesPerTenant()
                 || (currentMapSize == getConfig().getMaxDevicesPerTenant() && adapterInstancesForTenantMap.containsKey(deviceId))) {
-            adapterInstancesForTenantMap.put(deviceId,
-                    new ExpiringValue<>(createAdapterInstanceIdJson(protocolAdapterInstanceId), getLifespanNanos(lifespan)));
-            result = DeviceConnectionResult.from(HttpURLConnection.HTTP_NO_CONTENT);
+            final ExpiringValue<JsonObject> newValue = new ExpiringValue<>(
+                    createAdapterInstanceIdJson(protocolAdapterInstanceId), getLifespanNanos(lifespan));
+            if (updateOnly) {
+                final ExpiringValue<JsonObject> resultingValue = adapterInstancesForTenantMap
+                        .computeIfPresent(deviceId, (key, oldValue) -> {
+                            if (protocolAdapterInstanceId.equals(getAdapterInstanceIdFromJson(oldValue.getValue()))) {
+                                return newValue;
+                            } else {
+                                return oldValue;
+                            }
+                        });
+                if (resultingValue != null && protocolAdapterInstanceId.equals(getAdapterInstanceIdFromJson(resultingValue.getValue()))) {
+                    result = DeviceConnectionResult.from(HttpURLConnection.HTTP_NO_CONTENT);
+                } else {
+                    result = DeviceConnectionResult.from(HttpURLConnection.HTTP_PRECON_FAILED);
+                }
```

Comment:
I think we need to also remove an expired existing value if the adapter instance doesn't match ...

Reply:
@calohmn WDYT?

Reply:
Removing an expired entry here is something I don't quite understand - the removal of an expired entry is done under the hood by the Caffeine cache.

However, I've found two other issues here: one (updated lifespan wasn't taken into account) is already fixed (along with an added test), the other (a failed update influences lifespan of existing entry) is something I'm currently working on.

Reply:
But we are using a simple `Map` here for keeping the data, not a Caffeine Cache, or am I mistaken?

Reply:
The `adapterInstancesForTenantMap` is a Caffeine Cache under the hood (using `cache.asMap()`).

Reply:
I see

Reply:
2nd issue is fixed now.